cpp/src/arrow/dataset/scanner.h

```diff
@@ -39,27 +42,35 @@ struct ARROW_DS_EXPORT ScanContext {};
 class ARROW_DS_EXPORT ScanOptions {
  public:
   virtual ~ScanOptions() = default;
+
+  const DataSelector* selector() const { return selector_.get(); }
+
+ protected:
+  // Filters
+  std::unique_ptr<DataSelector> selector_;
```

Comment:
Thought: Selectors could be reused. No need to change anything until it becomes an issue though

Reply:
Changed to a shared_ptr. A notable difficulty with unique_ptrs is the pointed type must not be incomplete when it's instantiated. Some of CI was failing on that